File: go/promotions/stubpromotion.go
// ABOUTME: StubPromotion is a test stub for the Promotion interface.
// ABOUTME: Allows configurable behavior for testing the promotion engine.
package promotions

import (
	"promotionengine/domain"
	"promotionengine/pricing"
)

// StubPromotion is a test stub implementation of the Promotion interface.
// It allows you to configure its behavior for testing purposes.
type StubPromotion struct {
	PromoID         domain.PromotionID
	DiscountAmount  domain.Money
	ShouldApply     bool
	Target          string // "line" or "cart"
	DiscountDetails string
}

// ID returns the promotion identifier.
func (s *StubPromotion) ID() domain.PromotionID {
	return s.PromoID
}

// IsApplicable returns whether this promotion should apply.
func (s *StubPromotion) IsApplicable(cart domain.Cart, context pricing.PricingContext) bool {
	return s.ShouldApply
}

// Apply returns a configured discount if applicable.
func (s *StubPromotion) Apply(cart domain.Cart, context pricing.PricingContext) []pricing.AppliedDiscount {
	if !s.ShouldApply {
		return []pricing.AppliedDiscount{}
	}

	discount := pricing.NewAppliedDiscount(s.PromoID, s.DiscountAmount, s.Target, s.DiscountDetails)
	return []pricing.AppliedDiscount{discount}
}

File: typescript/src/promotions/StubPromotion.ts
// ABOUTME: Test helper that implements Promotion with configurable behavior
// ABOUTME: Used for testing the promotion engine without real promotion logic

import { Promotion } from './Promotion';
import { PromotionId } from '../domain/PromotionId';
import { Cart } from '../domain/Cart';
import { PricingContext, AppliedDiscount } from '../pricing';

export class StubPromotion implements Promotion {
  constructor(
    private readonly promotionId: PromotionId,
    private readonly shouldApply: boolean,
    private readonly discounts: AppliedDiscount[]
  ) {}

  id(): PromotionId {
    return this.promotionId;
  }

  isApplicable(_cart: Cart, _context: PricingContext): boolean {
    return this.shouldApply;
  }

  apply(_cart: Cart, _context: PricingContext): AppliedDiscount[] {
    return this.discounts;
  }
}
